File: backend/app/models/lost_found.py
"""Lost and found model for CampusHive."""

from app.database import db
from app.models.base import BaseModel, TimestampMixin


class LostFound(BaseModel, TimestampMixin):
    """lost_found table."""

    __tablename__ = 'lost_found'

    item_id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    organization_id = db.Column(
        db.Integer,
        db.ForeignKey('organizations.organization_id'),
        nullable=False,
    )
    posted_by = db.Column(
        db.Integer,
        db.ForeignKey('users.user_id'),
        nullable=False,
    )
    item_type = db.Column(db.Enum('LOST', 'FOUND'), nullable=False)
    title = db.Column(db.String(200), nullable=False)
    description = db.Column(db.Text, nullable=False)
    category = db.Column(db.String(100))
    location = db.Column(db.String(255))
    date_of_incident = db.Column(db.Date)
    resolved_at = db.Column(db.DateTime)
    status = db.Column(
        db.Enum('OPEN', 'CLOSED'),
        server_default=db.text("'OPEN'"),
    )

    __table_args__ = (
        db.Index('organization_id', 'organization_id'),
        db.Index('posted_by', 'posted_by'),
    )

    images = db.relationship('LostFoundImage', backref='item', lazy=True, cascade='all, delete-orphan')
    chat_messages = db.relationship('LostFoundChat', backref='item', lazy=True, cascade='all, delete-orphan')

    def to_dict(self):
        return {
            "item_id": self.item_id,
            "organization_id": self.organization_id,
            "posted_by": self.posted_by,
            "poster_name": self.posted_by_user.full_name if self.posted_by_user else None,
            "item_type": str(self.item_type),
            "title": self.title,
            "description": self.description,
            "category": self.category,
            "location": self.location,
            "date_of_incident": self.date_of_incident.isoformat() if self.date_of_incident else None,
            "status": str(self.status),
            "resolved_at": self.resolved_at.isoformat() if self.resolved_at else None,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
            "image_count": len(self.images) if self.images else 0,
            "images": [
                image.to_dict()
                for image in self.images
            ]
        }

    def __repr__(self):
        return f'<LostFound {self.item_id}>'

File: frontend/campushive_app/lib/models/lost_found.dart
class LostFound {
  const LostFound({
    required this.itemId,
    required this.organizationId,
    required this.postedBy,
    required this.posterName,
    required this.itemType,
    required this.title,
    required this.description,
    required this.category,
    required this.location,
    required this.dateOfIncident,
    required this.status,
    required this.resolvedAt,
    required this.createdAt,
    required this.updatedAt,
    required this.imageCount,
    required this.images,
  });

  final int itemId;
  final int organizationId;
  final int postedBy;
  final String? posterName;
  final String itemType;
  final String title;
  final String description;
  final String? category;
  final String? location;
  final DateTime? dateOfIncident;
  final String status;
  final DateTime? resolvedAt;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final int imageCount;
  final List<LostFoundImage> images;

  factory LostFound.fromJson(Map<String, dynamic> json) {
    final rawImages = json['images'];
    return LostFound(
      itemId: _int(json['item_id']),
      organizationId: _int(json['organization_id']),
      postedBy: _int(json['posted_by']),
      posterName: _stringOrNull(json['poster_name']),
      itemType: _string(json['item_type']),
      title: _string(json['title']),
      description: _string(json['description']),
      category: _stringOrNull(json['category']),
      location: _stringOrNull(json['location']),
      dateOfIncident: _date(json['date_of_incident']),
      status: _string(json['status']),
      resolvedAt: _date(json['resolved_at']),
      createdAt: _date(json['created_at']),
      updatedAt: _date(json['updated_at']),
      imageCount: _int(json['image_count']),
      images: rawImages is List
          ? rawImages
                .whereType<Map<String, dynamic>>()
                .map(LostFoundImage.fromJson)
                .toList()
          : const [],
    );
  }

  static int _int(Object? value) => value is num ? value.toInt() : 0;
  static String _string(Object? value) => value?.toString() ?? '';
  static String? _stringOrNull(Object? value) {
    final text = value?.toString().trim();
    return text == null || text.isEmpty ? null : text;
  }

  static DateTime? _date(Object? value) =>
      value is String ? DateTime.tryParse(value) : null;
}

class LostFoundImage {
  const LostFoundImage({required this.imageId, required this.imageUrl});

  final int imageId;
  final String imageUrl;

  factory LostFoundImage.fromJson(Map<String, dynamic> json) {
    return LostFoundImage(
      imageId: json['image_id'] is num ? (json['image_id'] as num).toInt() : 0,
      imageUrl: json['image_url']?.toString() ?? '',
    );
  }
}
